src/v1/entity/utils/validateKeyInput.ts

```diff
@@ -1,27 +1,138 @@
+
+  // Check that keyInput values match item or mapped
+  Object.entries(keyInput).forEach(([propertyName, propertyInput]) => {
+    const property = itemOrMapped._properties[propertyName]
+    // TODO, create joinPath util
+    const propertyPath = [path, propertyName].filter(Boolean).join('.')
+
+    if (property !== undefined) {
+      validateKeyInput(property, propertyInput, propertyPath)
+    } else {
+      if (isClosed(itemOrMapped)) {
+        throw new UnexpectedPropertyError({ path: propertyPath })
+      }
+      // TODO: create validateAny ?
+    }
+  })
+
+  // Check that all key & always required properties are present in keyInput
+  Object.entries(itemOrMapped._properties)
+    .filter(([, property]) => isKey(property) && property._required === 'always')
+    .forEach(([propertyName, property]) => {
+      const propertyKeyInput = keyInput[propertyName]
+      // TODO, create joinPath util
+      const propertyPath = [path, propertyName].filter(Boolean).join('.')
+
+      if (propertyKeyInput !== undefined) {
+        validateKeyInput(property, propertyKeyInput, propertyPath)
+      } else {
+        throw new MissingRequiredPropertyError({ path: propertyPath })
+      }
+    })
+}
 
 /**
  * Validates the primary key input of a single item command (GET, DELETE ...) for a given Entity
  *
- * @param entity Entity
+ * @param entry Entity | Item | Property
  * @param keyInput Key input
- * @return Boolean
+ * @param path _(optional)_ Path of the property in the related item (string)
+ * @return void
  */
-export const validateKeyInput: KeyInputValidator = <
-  E extends EntityV2,
-  K extends Record<string, any> = KeyInput<E>
->(
-  entity: E,
-  keyInput: Record<string, any>
-): keyInput is K => {
-  entity
-  keyInput
-  // TODO
-  return true
+export const validateKeyInput: KeyInputValidator = <E extends EntityV2 | Item | Property>(
+  entry: E,
+  keyInput: KeyInput<E>,
+  path?: string
+): void => {
+  if (entry._type === 'entity') {
+    return validateKeyInput(entry.item, keyInput)
+  }
+
+  if (entry._type === 'item') {
+    return validateProperties(entry, keyInput, path)
+  }
+
+  if (!isKey(entry)) throw new UnrecognizedKeyInputPropertyError({ path })
+
+  switch (entry._type) {
+    case 'any':
+      break
+    case 'binary':
+    case 'boolean':
+    case 'number':
+    case 'string':
+      const validator = validatorsByLeafType[entry._type]
+      if (!validator(keyInput))
+        throw new InvalidKeyInputValueTypeError({ expectedType: entry._type, keyInput, path })
+      break
+    case 'list':
+      if (!isArray(keyInput))
+        throw new InvalidKeyInputValueTypeError({ expectedType: 'list', keyInput, path })
+      keyInput.forEach((keyInputElement, index) =>
+        validateKeyInput(entry._elements, keyInputElement, `${path ?? ''}[${index}]`)
+      )
+      break
+    case 'map':
+      validateProperties(entry, keyInput, path)
+      break
+  }
```

Comment:
don't we also need to handle sets here?

Reply:
It's on the roadmap ^^ But sets are not implemented yet!